src/eval.rs

```diff
@@ -197,10 +285,147 @@ impl Expr {
+                    out.push(x.as_ref());
+                }
+                return out;
+            }
         }
         .into_iter()
         .map(|x| x.as_ref())
         .filter_map(Result::ok)
+        .map(|x| x.as_ref())
         .collect()
     }
+
+    pub fn get_definition(&self) -> Option<Gc<Expr>> {
+        use ExprSource::*;
+        match &self.source {
+            Ident { name } => self.scope.get(&name),
+            Select { from, index } => {
+                let idx = index.as_ref().ok()?.as_ident().ok()?;
+                let out = from
+                    .as_ref()
+                    .ok()?
+                    .eval()
+                    .ok()?
+                    .as_map()
+                    .ok()?
+                    .get(&idx)?
+                    .clone();
+                if let ExprSource::KeyValuePair { ref key, .. } = out.source {
+                    key.clone().ok()
+                } else {
+                    Some(out)
+                }
+            }
+            _ => None,
+        }
+    }
+
+    /// Interpret the expression as an identifier. For example:
+    /// ```text
+    /// foo => "foo"
+    /// "foo" => "foo"
+    /// "${"foo"}" => "foo"
+    /// ```
+    pub fn as_ident(&self) -> Result<String, EvalError> {
+        use ExprSource::*;
+        match &self.source {
+            Ident { name } => Ok(name.clone()),
+            Dynamic { inner } => inner.as_ref()?.eval()?.as_str(),
+            Literal { value } => value.as_str(),
+            _ => Err(EvalError::Internal(InternalError::Unimplemented(
+                "unsupported identifier expression".to_string(),
+            ))),
+        }
+    }
+}
+
+/// Used for merging sets during parsing. For example:
+/// { a.b = 1; a.c = 2; } => { a = { b = 1; c = 2; }; }
+pub fn merge_set_literal(name: String, a: Gc<Expr>, b: Gc<Expr>) -> Result<Gc<Expr>, EvalError> {
+    // evaluate literal attr sets only, otherwise error
+    let eval_literal = |src: Gc<Expr>| {
+        let src = if let ExprSource::KeyValuePair { value, .. } = &src.source {
+            value.as_ref()?.clone()
+        } else {
+            src
+        };
+        if let ExprSource::AttrSet { .. } = &src.source {
+            src.eval()?.as_map()
+        } else {
+            // We cannot merge a literal with a non-literal. This error is
+            // caused by incorrect expressions such as:
+            // ```
+            // repl> let x = { y = 1; }; in { a = x; a.z = 2; }
+            // error: attribute 'a.z' at (string):1:33 already defined at (string):1:26
+            // ```
+            Err(EvalError::Value(ValueError::AttrAlreadyDefined(name.to_string())))
+        }
+    };
+
+    let a = eval_literal(a)?;
+    let b = eval_literal(b)?;
+    let mut out = HashMap::new();
+    for (key, val) in a.iter() {
+        let tmp = match b.get(key) {
+            Some(x) => merge_set_literal(format!("{}.{}", name, key), x.clone(), val.clone())?,
```

Comment:
Please note that e.g. this works fine:

```
nix-repl> :p { b = { x = 1; }; b = { y = 2; }; }
{ b = { x = 1; y = 2; }; }
```

Reply:
I guess I was wrong, both

```
{
  b = { x = 1; };
  b = { y = 2; };
}.b.x
```

and

```
{
  b = { x = 1; };
  b = { y = 2; };
}.b.y
```

resolve the correct value, so this seems to be taken into account :+1: